src/v/cluster/rm_stm.cc

```diff
@@ -1886,10 +1886,36 @@ void rm_stm::apply_data(model::batch_identity bid, model::offset last_offset) {
     }
 }
 
-void rm_stm::apply_checkpoint([
-  [maybe_unused]] const model::record_batch& batch) {}
+void rm_stm::apply_checkpoint(
+  [[maybe_unused]] const model::record_batch& batch) {}
 
-ss::future<> rm_stm::checkpoint_in_memory_state() const { co_return; }
+ss::future<model::record_batch> rm_stm::make_checkpoint_batch() {
+    iobuf key;
+    co_await serde::write_async(key, _insync_term);
+    iobuf val;
+    co_await serde::write_async(val, std::move(_mem_state));
+    storage::record_batch_builder builder(
+      model::record_batch_type::tx_checkpoint, model::offset(0));
+    builder.set_control_type();
+    builder.add_raw_kv(std::move(key), std::move(val));
+    co_return std::move(builder).build();
+}
+
+ss::future<> rm_stm::checkpoint_in_memory_state() {
+    auto batch = co_await make_checkpoint_batch();
+    _mem_state = {}; // reset
```

Comment:
it seems like it might be easier to understand if the calling context explicitly took ownership of _mem_state content and passed it into checkpointing routines so that clearing that structure isn't a side effect of data serialization / replication?